app/controllers/users/phone_setup_controller.rb

```diff
@@ -135,5 +139,30 @@ def new_phone_form_params
         :recaptcha_mock_score,
       )
     end
+
+    def check_phone_submission_limit(phone)
+      return false if phone.blank?
+      fingerprint = Pii::Fingerprinter.fingerprint(Phonelib.parse(phone).e164.to_s)
+      @submission_rate_limiter ||= RateLimiter.new(
+        target: fingerprint,
+        rate_limit_type: :phone_fingerprint_confirmations,
+      )
+      @submission_rate_limiter.increment!
+      lock_out_user if @submission_rate_limiter.maxed?
+    end
+
+    def lock_out_user(user: current_user)
+      UpdateUser.new(user: user, attributes: { second_factor_locked_at: Time.zone.now }).call
+      flash[:error] = t(
+        'errors.messages.phone_confirmation_limited',
+        timeout: distance_of_time_in_words(
+          Time.zone.now,
+          [@submission_rate_limiter.expires_at, Time.zone.now].compact.max,
+          except: :seconds,
+        ),
+      )
+      redirect_to account_path
```

Comment:
I really don't think we should redirect the user to the account page, since that may throw them off course if they had intended to return to the service provider.

This kind of validation should probably be part of `NewPhoneForm`, to keep the controller [lean](https://github.com/18F/identity-idp/blob/main/docs/backend.md#forms-formresponse-analytics-and-controllers) and so that it can take advantage of the existing unsuccessful submission handling in the `create` method above.

That being said, `create` doesn't handle error messages, or at least it didn't as of the current branch point of this pull request. This was [changed](https://github.com/18F/identity-idp/blob/f441db9c64bdced4200d75e0c5a3d0946b52cd1e/app/controllers/users/phone_setup_controller.rb#L39) in #10454 to support error messages for reCAPTCHA. I could maybe imagine the code for `first_error_message` being updated to support multiple keys so that we could surface validation errors for _either_ reCAPTCHA or rate-limiting.

```rb
flash.now[:error] = result.first_error_message(:rate_limited, :recaptcha_token)
```

At the very least, I'd think we should redirect back to the phone setup form so the error message is shown on the same page, not redirect the user away..

Reply:
That makes sense to let the limiter do the work. I have changed that and moved the validation to the form. I'm not 100% sure how I executed the error message since it doesn't use flash but it looks like it's working.